### DISCUSSION TITLE
Style editing stumpwm to performance

### DISCUSSION THREAD

**Discussion by equwal** [stumpwm/stumpwm#1120] [General]
I'm just going to list style edits here that can be done. They should be straightforward and have the potential for improving performance, although only if the style is improved too. For performance uglification profiling must be done.

Super obvious ones to me:

`(> (length list) 1)`  to `(not (null list))`. O(n) list grab. I like not null because it says what it does more than just `list`.
`(format nil *variable* stuff)` -> `format nil +string-constant+` or just put the string constant there. Where reasonable.

**Comment by equwal** [comment]
Groups are done by passing around a lisp. And they are slow and have always been slow. At least it would be nice to see destructive idioms instead of the obvious function copying function. I've seen reverse used a fair bit, why not use the destructive idiom instead, I think it was push/nreverse? The compiler can't do it for you.

**Comment by equwal** [comment]
https://github.com/stumpwm/stumpwm/issues/382 https://github.com/stumpwm/stumpwm/issues/1039

**Comment by dmb2** [comment]
I would like to see:

- Demonstrable evidence that one approach is better than another in terms of performance.
   - I would like to see specific proof of a performance difference between `(> (length list) 1)` and `(not (null list))`.
- Actual measurements that show a specific stylistic change positively affects performance.

**Comment by equwal** [comment]
The point of this approach is that a lot of obvious edits can be made without having to profile everything. Which is why it should only be done when it improves style too.

Whether it matters depends on the length of the list. `consp list` is better style than `go through the entire list to find out if it has an element`.

```
(let ((big-boy (loop for i from 0 to 1000000 collect i)))
  (time (length big-boy))
  (time (cdr big-boy)))
```
```
;Evaluation took:
;  0.040 seconds of real time
;  0.041104 seconds of total run time (0.041104 user, 0.000000 system)
;  102.50% CPU
;  121,960,828 processor cycles
;  0 bytes consed
;  
;Evaluation took:
;  0.000 seconds of real time
;  0.000003 seconds of total run time (0.000003 user, 0.000000 system)
;  100.00% CPU
;  1,265 processor cycles
;  0 bytes consed
```

**Comment by equwal** [comment]
It's a lot of extra work to profile low-level utilities and the inside of a lot of loops, just to see if it is worth doing `(let ((var (length list)) (loop ...var...))`. When it is worth doing period.

**Comment by dmb2** [comment]
Your example is contrived to make your point and doesn't match the code you're proposing to change.  You need to time the code that you are actually going to run. Compilers are smart pieces of software and SBCL is no different:
```
(let ((big-boy (loop for i from 0 to 1000000)))
  (time (> (length big-boy) 1))
 (time (not (null big-boy))))
Evaluation took:
  0.000 seconds of real time
  0.000001 seconds of total run time (0.000001 user, 0.000000 system)
  100.00% CPU
  320 processor cycles
  0 bytes consed
  
Evaluation took:
  0.000 seconds of real time
  0.000001 seconds of total run time (0.000001 user, 0.000000 system)
  100.00% CPU
  114 processor cycles
  0 bytes consed
  
NIL
```
I personally am not going to merge or support changing code (at the risk of introducing new bugs and regressions) to save 206 *processor cycles*

**Comment by equwal** [comment]
there was a bug in my contrived example and I edited it

**Comment by dmb2** [comment]
There is a difference when the for loop actually returns a list: 
```
(let ((big-boy (loop for i from 0 to 1000000 collect i)))
  (time (> (length big-boy) 1))
 (time (not (null big-boy))))
Evaluation took:
  0.000 seconds of real time
  0.001370 seconds of total run time (0.001268 user, 0.000102 system)
  100.00% CPU
  3,548,130 processor cycles
  0 bytes consed
  
Evaluation took:
  0.000 seconds of real time
  0.000001 seconds of total run time (0.000001 user, 0.000000 system)
  100.00% CPU
  192 processor cycles
  0 bytes consed
```
I still don't like the idea of changing existing code for the following reasons:
1. The codebase is quite stable, and stylistic changes have a tendency to introduce more bugs than they fix.
2. The performance issues you point to in the linked issues are probably not coming from they way we check whether an array has non-zero length.
3. In this specific example, the lengths of the lists involved will not be in the millions.

If you are interested in improving performance, we need good ways to profile the code and quantitative evidence that a new method is significantly better and does not introduce regressions.

Yes, that is a lot of work.  Furthermore, performance improvements will likely come from restructuring how we do things, not from micro optimizations that have the potential to introduce additional bugs do to subtle differences in the semantics.

**Reply by equwal** [reply]
It can get into the millions when it is inside of a loop, inside of a function called by another function inside of a loop etc.

**Reply by equwal** [reply]
I understand your point of view, but I think bugs can be avoided by using a refactoring program to make the edits, and being careful to only make edits that don't actually change anything.

**Comment by equwal** [comment]
I understand your point of view. Since it isn't entirely necessary to do what I suggest, I'll close the discussion.
